### ISSUE TITLE
Curried Functions Being Reduced Will throw positional argument error

### ISSUE THREAD

**Issue by mtbsickrider** [dgilland/pydash#112]
```python
import pydash as _
@_.curry
def curried_function(data, accum, id):
    accum[id] = _.reduce_(data, sum)
    return accum

ids = [1]
data = [1,2]
curried_function_with_data = curried_function(data)
_.reduce_(ids, curried_function_with_data, {}) #throws Error since it tries to callit with index since argscount = 3.
```

Can the function getargcount subtract from the amount of args the function already has?

Also im very interested in getting started in open source, so wouldn't mind contributing if you think this might be useful.

**Comment by dgilland** [comment]
The problem is that `getargcount()` just looks at the method signature and if it sees `varargs` then it assumes that all variables can be passed to the callback.

A possible fix would be to use the arity value in the curried function as an arg limit when calling the wrapped function. Something like the following inserted into the `Curry.__call__` method:

```python
if (len(args) + len(kargs)) >= self.arity:
    args_arity = self.arity - len(kargs)
    args = args[:(args_arity if args_arity > 0 else 0)]
    curried = self.func(*args, **kargs)
```

With that patch and a slight modification to your example (the `_.reduce_(data, sum)` call won't work since `sum` expects an iterable but `_.reduce_()` will do `callback(result, item, index)`):

```python
import pydash as _

@_.curry
def curried_function(data, accum, id):
    accum[id] = _.reduce_(data, lambda total, n: total + n)
    return accum

ids = [1]
data = [1,2]
curried_function_with_data = curried_function(data)
result = _.reduce_(ids, curried_function_with_data, {})
print(result)
# {1: 3}
```

But if you are just wanting to sum the values in `data`, then you may not need to use `reduce_` and could just do `accum[id] = sum(data)` but I may be missing something.

**Comment by mtbsickrider** [comment]
Oh,the code presented is a simplified case of what im really doing.  The goal was to identify what to do regarding functions that are passed into reduce that have been curried.

**Comment by mtbsickrider** [comment]
The actual code: 

```
@_.curry
def get_data_by_id(fields, accum, obj, index):
    id = obj.get('id')
    accum[id] = {k: _.get(obj, v, '') for k, v in fields.items()}
    return accum
```

```
    feature_fields = dict(
        id='id',
        danceability='danceability',
        energy='energy',
        key='key',
        loudness='loudness',
        speechiness='speechiness',
        acousticness='acousticness',
        instrumentalness='instrumentalness',
        valence='valence',
        tempo='tempo',
    )
    get_data_by_id_features = get_data_by_id(feature_fields)
    features_data = _.reduce_(features, get_data_by_id_features, {})
```

This code is my current workaround, but i dont need to use the index. Removing the index parameter then would throw arg error.

**Comment by dgilland** [comment]
I believe the patch I mentioned above will fix things so that you wouldn't have to have the `index` argument in the function.

What I'm proposing is to replace what's on this line:

https://github.com/dgilland/pydash/blob/7fa58c183397be5e6edf3677ce21833f05cd9933/pydash/functions.py#L145

with:

```python
if (len(args) + len(kargs)) >= self.arity:
    args_arity = self.arity - len(kargs)
    args = args[:(args_arity if args_arity > 0 else 0)]
    curried = self.func(*args, **kargs)
```

**Comment by dgilland** [comment]
FYI, this has been fixed in `v4.3.2`: https://pypi.python.org/pypi/pydash/4.3.2
